graphql/redact_secrets_plugin.go

```diff
@@ -0,0 +1,117 @@
+package graphql
+
+import (
+	"fmt"
+	"os"
+	"sort"
+
+	"github.com/99designs/gqlgen/codegen/config"
+)
+
+// GenerateSecretFields generates a file that contains a list of fields that should be redacted in logs.
+func GenerateSecretFields(cfg *config.Config) error {
+	redactedFields := make(map[string]bool)
+
+	for _, schemaType := range cfg.Schema.Types {
+		for _, field := range schemaType.Fields {
+			if field.Directives.ForName("redactSecrets") != nil {
+				redactedFields[field.Name] = true
+			}
+		}
+	}
+
+	var fields []string
+	for field := range redactedFields {
+		fields = append(fields, field)
+	}
+
+	// Sort the fields to ensure consistent output.
+	sort.Sort(sort.StringSlice(fields))
+
+	return generateRedactedFieldsFile(fields)
+}
+
+// generateRedactedFieldsFile generates a file that contains a list of fields that should be redacted in logs. It is used to generate the redacted_fields_gen.go file.
+// The file contains a list of fields that should be redacted in logs, a function to check if a field should be redacted, and a function to redact fields in a map.
+func generateRedactedFieldsFile(fields []string) error {
+	file, err := os.Create("graphql/redacted_fields_gen.go")
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	_, err = file.WriteString(`// Code generated by graphql/redact_secrets_plugin.go DO NOT EDIT.
+package graphql
+
+import (
+	"reflect"
+)
+
+var RedactedFields = []string{
+`)
+	if err != nil {
+		return err
+	}
+
+	for _, field := range fields {
+		_, err = file.WriteString(fmt.Sprintf("\t\"%s\",\n", field))
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = file.WriteString("}\n")
+	if err != nil {
+		return err
+	}
+
+	_, err = file.WriteString(
+		`
+func IsFieldRedacted(fieldName string) bool {
+	for _, field := range RedactedFields {
+		if field == fieldName {
+			return true
+		}
+	}
+	return false
+}
+
+// RedactFieldsInMap recursively searches for and redacts fields in a map.
+// Assumes map structure like map[string]interface{} where interface{} can be another map, a slice, or a basic datatype.
```

Comment:
Can we do some input validation to assert this assumption?

Reply:
Can you clarify what you mean by this? The `reflect.TypeOf(value).kind()` statements perform the validation to validate the type of fields in the interface.

Reply:
Right - that should be sufficient, resolving